Configure OCIO for xstudio

philippe-ynput opened the pull request:
## Changelog Description

Allow the loader to fill the OCIO environment variable for the current representation.

Fixes #3 

## Additional review information

Implement an ad-hoc solution to find relevant data in the context passed to the loader.

## Testing notes:
1. start with this step
2. follow this step

philippe-ynput committed:
Update windows path to xstudio

Signed-off-by: Philippe Leprince <philippe@ynput.io>

philippe-ynput committed:
setup OCIO for the current representation

Signed-off-by: Philippe Leprince <philippe@ynput.io>

jakubjezek001 commented on client/ayon_xstudio/plugins/load/open_in_xstudio.py:
```diff
@@ -110,13 +112,60 @@ def load(
+        )
+        log.info("anatomy.roots = %s", anatomy.roots)
+        log.warning("Not configuring OCIO !")
+        return
```
We should have also try if the path exists before we return. In my case the path is not de-rootable but exists, but the process is ignoring it.

iLLiCiTiT commented on client/ayon_xstudio/plugins/load/open_in_xstudio.py:
```diff
@@ -110,13 +112,60 @@ def load(
         ]
 
+        # TODO(plp): Is there a get_ayon_env() somewhere ?
+        ayon_env = _set_ocio_env_var(context, dict(os.environ))
```
```suggestion
        ayon_env = dict(os.environ)
        _set_ocio_env_var(context, ayon_env)
```

iLLiCiTiT commented on client/ayon_xstudio/plugins/load/open_in_xstudio.py:
```diff
@@ -110,13 +112,60 @@ def load(
+        ocio_path = ocio_path.resolve()
+
+    anatomy = Anatomy(context["project"]["name"])
+    ok, rootless_path = anatomy.find_root_template_from_path(
```
Isn't the ocio path with root template? `{root[work]}`?

BTW we should implement some kind of function in ayon-core to receive ocio path from representation...

jakubjezek001 replied:
yes it would be great, and even de-rootify the abs paths by default

jakubjezek001 commented on client/ayon_xstudio/plugins/load/open_in_xstudio.py:
```diff
@@ -94,9 +98,7 @@ def load(
             # we output the seq path with hashes.
             idxs = list(seq.indexes)
             pad = "#" * (seq.padding + len(str(idxs[0])))
```
```suggestion
            pad = "#" * seq.padding
```

Here is what was happeing to me on some sequences:
```
path : C:\projects\AY01_VFX_demo\shots\sq060\sh0010\publish\render\renderCompositingMain\v001\a01vfxd_sh0010_renderCompositingMain_v001.0999.exr  ]
>>> [  pattern : \.(?P<index>(?P<padding>0*)\d+)\.\D+\d?$  ]
>>> [  collections : [<Collection "a01vfxd_sh0010_renderCompositingMain_v001.%04d.exr [999-1007]">]  ]
>>> [  1first_image : None  ]
>>> [  idxs : [999, 1000, 1001, 1002, 1003, 1004, 1005, 1006, 1007]  ]
>>> [  seq.padding : 4  ]
>>> [  pad : #######  ]
>>> [  2first_image : a01vfxd_sh0010_renderCompositingMain_v001.#######.exr=999-1007  ]
>>> [  Opening xStudio with : C:\projects\AY01_VFX_demo\shots\sq060\sh0010\publish\render\renderCompositingMain\v001\a01vfxd_sh0010_renderCompositingMain_v001.#######.exr=999-1007  ]
```

iLLiCiTiT replied:
I think it was meant as this...
```python
            pad = "#" * max(seq.padding + len(str(idxs[-1])))
```

But if it just expects to mark the frames "location in the string" then Jeza's comment would do the work too.

jakubjezek001 commented on client/ayon_xstudio/plugins/load/open_in_xstudio.py:
```diff
@@ -94,9 +98,7 @@ def load(
             # we output the seq path with hashes.
             idxs = list(seq.indexes)
             pad = "#" * (seq.padding + len(str(idxs[0])))
```
```suggestion
            pad = "#" * len(str(idxs[-1]))
```

Alternatively safer is to use this, since some versions of clique was returning on seq.padding `0`. Assuming this might be history - just mentioning that this used to happen to me and this was workaround.

philippe-ynput committed:
fix(xstudio): address review comments.

- Use last index length for sequence padding instead of padding + first index length
- Refactor OCIO environment setup to modify environment dict in-place
- Add explicit path existence check for OCIO config validation

Signed-off-by: Philippe Leprince <philippe@ynput.io>

philippe-ynput commented:
@iLLiCiTiT @jakubjezek001 There is still something inherently unsatisfactory in this code:
- The representation's OCIO path points to Libor's Nuke config.
- I don't have Nuke installed so the derootification failed, but I could have that same OCIO config available through `ayon-ocio`.
How can I handle that case ?

iLLiCiTiT commented:
> I don't have Nuke installed so the derootification failed, but I could have that same OCIO config available through ayon-ocio.
How can I handle that case ?

Related to built-in ocio, there is opened issue for that, which is unsolved https://github.com/ynput/ayon-core/issues/969 .

philippe-ynput commented:
Thanks @iLLiCiTiT 
So I could:
- get the ocio config settings
- identify the original representation config
- pass a matching config with maybe a warning ?

iLLiCiTiT commented:
All possible options are potentially dangerous. I think the safest would be to check if the path contains `"ayon_ocio"` and parse the built in ocio name from that? Maybe?

philippe-ynput committed:
feat(xstudio): add OCIO config resolution fallback

Enhance OCIO environment variable setup by adding a fallback mechanism to find AYON OCIO config when the initial path does not exist. This improves compatibility and ensures proper color management for XStudio integrations by resolving paths through anatomy templates or alternative config discovery.

Signed-off-by: Philippe Leprince <philippe@ynput.io>

philippe-ynput committed:
style(ruff): minor fixes.

Signed-off-by: Philippe Leprince <philippe@ynput.io>

philippe-ynput commented:
@iLLiCiTiT @jakubjezek001 Please review.

iLLiCiTiT commented on client/ayon_xstudio/plugins/load/open_in_xstudio.py:
```diff
@@ -93,10 +97,8 @@ def load(
-            first_image = (
-                f"{seq.head}{pad}{seq.tail}={idxs[0]}-{idxs[-1]}"
-            )
+            pad = "#" * len(str(idxs[-1]))
```
Maybe it was already discussed but isn't this causing possible issue? If I would have frames 1-10, then this would be `##`, but the frame string in filename is `0010`, so it should be `####`. Which I belive the previous logic did handle.

iLLiCiTiT commented on client/ayon_xstudio/plugins/load/open_in_xstudio.py:
```diff
@@ -110,13 +112,106 @@ def load(
+    representation: dict = context.get("representation", {})
+    if not representation:
+        err = "Couldn't find 'representation' in context !"
+        raise KeyError(err)
```
You trigger this only from the load plugin which 100% have representation otherwise it would not trigger `load`.
```suggestion
    representation: dict = context["representation"]
```

iLLiCiTiT commented on client/ayon_xstudio/plugins/load/open_in_xstudio.py:
```diff
@@ -110,13 +112,106 @@ def load(
+        ocio_path = ocio_path.resolve()
+
+    if not ocio_path.exists():
+        anatomy = Anatomy(context["project"]["name"])
```
You can pass in the project from context
```suggestion
        anatomy = Anatomy(
            context["project"]["name"],
            project_entity=context["project"],
        )
```

iLLiCiTiT commented on client/ayon_xstudio/plugins/load/open_in_xstudio.py:
```diff
@@ -110,13 +112,106 @@ def load(
+        ocio_folder = get_ocio_config_path()
+        log.debug("ocio root: %s", ocio_folder)
+        folder_index = path.parts.index("OpenColorIOConfigs")
+        server_ocio = Path(ocio_folder).joinpath(
```
This needs some comment, because I have no idea what and why is happening here.

iLLiCiTiT replied:
Ok, I do understand. But I think we should change the order of things that do happen here. Also the reassignment of `path = Path(ocio_path)` is not really needed, or?
```python
    log.debug("representation ocio: %s", ocio_path.as_posix())

    # Try to import ayon_ocio to get builtin ocio directory.
    try:
        from ayon_ocio import get_ocio_config_path
    except ImportError:
        log.debug("Failed to import 'get_ocio_config_path'")
        return None
    
    if "OpenColorIOConfigs" not in ocio_path.parts:
        return None

    folder_index = path.parts.index("OpenColorIOConfigs")

    ayon_ocio_dir = get_ocio_config_path()
    log.debug("ocio root: %s", ayon_ocio_dir)

    server_ocio = Path(ayon_ocio_dir).joinpath(
        *ocio_path.parts[folder_index + 1:]
    )
    log.debug("server_ocio = %s", server_ocio)
    if not server_ocio.exists():
        log.debug("server_ocio doesn't exist !")
        return None
    return server_ocio
```

iLLiCiTiT commented on client/ayon_xstudio/plugins/load/open_in_xstudio.py:
```diff
@@ -71,6 +74,7 @@ def load(
             namespace (Optional[str]): The namespace of the item to load.
             options (Optional[Any]): Additional options for loading.
         """
+        # print(context)
```
```suggestion
```

jakubjezek001 reviewed (approved):
it was working well

iLLiCiTiT committed:
Merge branch 'develop' into OCIO-config

# Conflicts:
#	client/ayon_xstudio/plugins/load/open_in_xstudio.py
